crates/sui-core/src/unit_tests/gas_tests.rs

```diff
@@ -229,6 +230,66 @@ async fn test_transfer_sui_insufficient_gas() {
     assert_eq!(effects.mutated()[0].1, sender);
 }
 
+fn generate_random_gas_data(sender: SuiAddress) -> (GasData, Vec<Object>) {
+    let mut rng = rand::thread_rng();
+    let mut gas_objects = vec![];
+    let mut object_refs = vec![];
+
+    // Limits set by the protocol. We want to generate random gas data that's within these limits.
+    let max_gas_balance = TOTAL_SUPPLY_MIST;
+    let max_num_gas_objects = ProtocolConfig::get_for_max_version().max_gas_payment_objects();
+    let max_gas_budget = ProtocolConfig::get_for_max_version().max_tx_gas();
+
+    let total_gas_balance = rng.gen_range(0..=max_gas_balance);
+    let mut remaining_gas_balance = total_gas_balance;
+    let num_gas_objects = rng.gen_range(1..=max_num_gas_objects); // need at least one gas object
+    for _ in 0..num_gas_objects - 1 {
+        let gas_balance = rng.gen_range(0..=remaining_gas_balance);
+        // TODO: generate gas obj with arbitrary owner too.
+        let gas_object = Object::new_gas_with_balance_and_owner_for_testing(gas_balance, sender);
+        remaining_gas_balance -= gas_balance;
+        object_refs.push(gas_object.compute_object_reference());
+        gas_objects.push(gas_object);
+    }
+    // Put the remaining balance in the last gas object.
+    gas_objects.push(Object::new_gas_with_balance_and_owner_for_testing(
+        remaining_gas_balance,
+        sender,
+    ));
+
+    assert_eq!(gas_objects.len(), num_gas_objects as usize);
+    assert_eq!(gas_objects.iter().map(|o| o.data.try_as_move().unwrap().get_coin_value_unsafe()).sum::<u64>(), total_gas_balance);
+
+    (GasData {
+        payment: object_refs,
+        owner: sender,
+        price: rng.gen_range(0..=u64::MAX),
+        budget: rng.gen_range(0..=std::cmp::min(total_gas_balance, max_gas_budget)),
```

Comment:
we might just want to let this one be `u64::MAX` as well--will test insufficient gas coins for budget, possible overflow in budget, etc.